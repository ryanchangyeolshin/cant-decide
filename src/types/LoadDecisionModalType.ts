import DecisionDataType from './DecisionDataType';

type LoadDecisionModalType = {
  open: boolean,
  handleClose: (event: React.MouseEvent<HTMLElement>, modalType: string) => void,
  decisions: DecisionDataType[],
  loading: boolean
};

export default LoadDecisionModalType;