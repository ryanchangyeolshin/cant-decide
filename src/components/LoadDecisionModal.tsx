import React, { useContext, useState } from 'react';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
import { Modal, Backdrop, Fade, Button, CircularProgress, List } from '@material-ui/core';
import LoadedDecisionItem from './LoadedDecisionItem';
import LoadDecisionModalType from '../types/LoadDecisionModalType';
import DecisionContext from '../context/DecisionContext';
import DecisionDataType from '../types/DecisionDataType';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    modal: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    paper: {
      backgroundColor: theme.palette.background.paper,
      border: '2px solid #000',
      boxShadow: theme.shadows[5],
      padding: theme.spacing(2, 4, 3),
    },
    modalScreen: {
      background: "#154A40",
      display: "flex",
      flexDirection: "column",
      justifyContent: "space-between",
      padding: "30px",
      width: "80%",
      height: "80%",
      "& .MuiList-root": {
        height: "100%",
        overflow: "scroll",
      }
    },
    loadDecision: {
      backgroundColor: "#F4ECDA",
      color: "black",
      marginTop: "40px",
      marginBottom: "10px",
      "&:hover": {
        backgroundColor: "yellow"
      },
      "& .MuiButton-label": {
        fontWeight: 900
      }
    }
  }),
);

const LoadDecisionModal: React.FC<LoadDecisionModalType> = ({ open, handleClose, decisions, loading }: LoadDecisionModalType) => {
  const { setDecision } = useContext(DecisionContext);
  const [selectedDecision, setSelectedDecision] = useState<DecisionDataType>({ _id: "", user: "", name: "", choices: [] });
  const classes = useStyles();

  const handleCloseModalHandler = (e: React.MouseEvent<HTMLElement>) => {
    setSelectedDecision({ _id: "", user: "", name: "", choices: [] });
    handleClose(e, "load");
  }

  const setSelectedDecisionHandler = (e: React.MouseEvent<HTMLElement>, decision: DecisionDataType) => {
    setSelectedDecision(decision);
  };

  const loadDecisionHandler: React.MouseEventHandler<HTMLButtonElement> = (e: React.MouseEvent<HTMLButtonElement>) => {
    setDecision(selectedDecision);
    handleCloseModalHandler(e);
  };

  return (
    <Modal
      id="load-decision-modal"
      aria-labelledby="transition-modal-title"
      aria-describedby="transition-modal-description"
      className={classes.modal}
      open={open}
      onClose={handleCloseModalHandler}
      closeAfterTransition
      BackdropComponent={Backdrop}
      BackdropProps={{
        timeout: 500,
      }}
    >
      {
        loading ? (
          <CircularProgress />
        ) : (
        <Fade in={open}>
          <div className={classes.modalScreen}>
            <List>
              {decisions.map(savedDecision => (
                <LoadedDecisionItem
                  key={savedDecision._id}
                  decision={savedDecision}
                  setSelectedDecisionHandler={setSelectedDecisionHandler}
                />
              ))}
            </List>
            <Button
              id="load-decision-button"
              className={classes.loadDecision}
              variant="contained"
              color="primary"
              disabled={selectedDecision._id === ""}
              onClick={loadDecisionHandler}
            >
              Load Decision
            </Button>
          </div>
        </Fade>
        )
      }
    </Modal>
  );
}

export default LoadDecisionModal;