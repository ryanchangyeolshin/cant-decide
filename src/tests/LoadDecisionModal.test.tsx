import React from 'react';
import { render, cleanup } from '@testing-library/react';
import LoadDecisionModal from '../components/LoadDecisionModal';
import LoadDecisionModalType from '../types/LoadDecisionModalType';

afterEach(cleanup);

jest.useFakeTimers();

describe("This suit is to test the LoadDecisionModal component", () => {
  const props: LoadDecisionModalType = {
    open: true,
    handleClose: (event: React.MouseEvent<HTMLElement>, modalType: string) => "closed the modal",
    decisions: [{ _id: "192e8du1id", name: "What to eat?!", user: "johndoe@gmail.com", choices: ["Taco Bell", "KFC"] }],
    loading: false
  };

  test('Snapshot of LoadDecisionModal', () => {
    const { asFragment } = render(
      <LoadDecisionModal
        open={true}
        handleClose={props.handleClose}
        decisions={props.decisions}
        loading={props.loading}
      />
    );
    jest.advanceTimersByTime(30000);
    expect(asFragment()).toMatchSnapshot()
  });
});
