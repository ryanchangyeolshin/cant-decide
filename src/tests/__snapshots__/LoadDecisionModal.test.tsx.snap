// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`This suit is to test the LoadDecisionModal component Snapshot of LoadDecisionModal 1`] = `<DocumentFragment />`;
